crates/typst/src/layout/grid/rowspans.rs

```diff
@@ -0,0 +1,533 @@
+            // requested by rowspans was properly covered, since combining the
+            // already grown amount with the total height contributed by future
+            // rows gives us a number that is larger than or equal to the total
+            // size requested by rowspans, and thus we stop, as we found how
+            // much we have to grow by.
+            //
+            // A flaw of this approach is that we consider rowspans' content to
+            // be contiguous. That is, we treat rowspans' requested heights as
+            // a simple number, instead of properly using the vector of
+            // requested heights in each region. This can lead to some
+            // weirdness when using multi-page rowspans with content that
+            // reacts to the amount of space available, including paragraphs.
+            // However, this is probably the best we can do for now.
+            let mut extra_amount_to_grow =
+                max_growable_height - amount_to_grow - total_spanned_height;
+            if extra_amount_to_grow <= Abs::zero() {
+                // The amount to grow is enough to fully cover the rowspans.
+                // Reduce sizes by the amount actually spanned by gutter.
+                subtract_end_sizes(
+                    &mut simulated_sizes,
+                    max_growable_height - amount_to_grow,
+                );
+                if let Some(last_resolved_size) = last_resolved_size {
+                    // Ensure the first simulated size is at least as large as
+                    // the last resolved size (its initial value). As it was
+                    // already resolved before, we must not reduce below the
+                    // resolved size to avoid problems with non-rowspan cells.
+                    if let Some(first_simulated_size) = simulated_sizes.first_mut() {
+                        first_simulated_size.set_max(last_resolved_size);
+                    } else {
+                        simulated_sizes.push(last_resolved_size);
+                    }
+                }
+                resolved.extend(simulated_sizes);
+                return Ok(());
+            }
+
+            // The amount to grow the auto row by has changed since the last
+            // simulation, but might not be enough yet to fully cover the
+            // rowspans. Let's try again or abort if we reached the max
+            // attempts.
+            amount_to_grow += extra_amount_to_grow;
+
+            // For the next simulation attempt, we consider that the auto row
+            // has additionally grown by the amount given in this attempt, to
+            // see if it will have to grow further in the next attempt.
+            while !extra_amount_to_grow.is_zero()
+                && simulated_regions.size.y < extra_amount_to_grow
+            {
+                extra_amount_to_grow -= regions.size.y.max(Abs::zero());
+                regions.next();
+            }
```

Comment:
I think I understand what this is supposed to do, but don't understand why this code does it. Could you explain it?

Reply:
I've fixed the bug now, it correctly advances `simulated_regions`.